add `convexHull`

andreas-web opened the pull request:
This PR adds the function `convexHull` and an helper function for JXG.Coords.

This functionality is required by sketchometry. In addition, intelligent label positioning will probably require this function in the future.

alfredwassermann reviewed (commented):
JSXGraph contains already the convex hull algorithm `JXG.Math.Geometry.GrahamScan()`.  Do we need another one? Maybe, a wrapper function would be sufficient.

andreas-web commented:
I did not know that there has already been a function. I just tested GrahamScan in sketchometry and it seems to lead to errors sometimes. Maybe because of the fact that GrahamScan always returns an array of coordsArrays. I would expect to get an array of Coords if I passed an array of Coords as parameter.

andreas-web commented:
I have now added a wrapper function `convexHull` and renamed my original function to `MonotoneChain`. It now also returns coords arrays

The wrapper function uses `GrahamScan` as default. After execution, the result is converted back to the original type (points, coordinates or arrays).

However, even if GrahamScan and MonononeChain seem to work the same (see https://activities.tjhsst.edu/sct/lectures/1415/SCT_Convex_Hull.pdf), I still need MonononeChain, as it often results in a different number of points. Maybe you can review the functions and tell me where my mistake lies.

alfredwassermann commented:
Thank you very much for the PR!

Indeed, `GrahamScan` was broken since a couple of years. The regression had slipped in with the clipping algorithms. Now it is fixed and refactored. In parallel, I had commited a method `convexHull` with slightly different return values from yours:

- If called as `JXG.Math.Geometry.convecHull(points)` the convex hull is returned as an array of pointers to the input elements, regardless of the type of the input elements (the input array may contain a mixture of types).
- If called as `JXG.Math.Geometry.convecHull(points)`  an array of coords arrays is returned, see the examples in the API docs of 554b3ec1596c5ddc9e31d20d04c18bd6b87bb494. This makes it relatively convenient to do a dynamic plot of the convex hull.

For this, the array returned by `GrahamScan` now consists of objects containing the index of the original point and a coords array.

Regarding `MonotoneChain`: could the returned array changed accordingly? Moreover, it should use `Array.sort` instead of bubble sort and use `JXG.Math.Geometry.signedTriangle` , which is `cross` divided by 2.

Best wishes,
Alfred
